{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c86dd8b1-e9b8-4859-a535-88625cef75b5",
   "metadata": {},
   "source": [
    "# Dataset Preparation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b61f9842-aee0-4e3a-ac81-bd3984c60c28",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "997a4165-d844-439e-94fe-0fdc34caf225",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/hnasrulloh/Works/pangenomes/notebooks'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "08c77550-4653-4910-a997-3b47a748615a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict, namedtuple\n",
    "from os import path, listdir, makedirs\n",
    "import re\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "55a3e6ca-5667-4d4e-a5f6-957756c3edee",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_DIR = path.join(path.pardir, 'data')\n",
    "\n",
    "PATRIC_AMR = path.join(DATA_DIR, 'PATRIC_genomes_AMR.txt')\n",
    "GENOME_SUMMARY = path.join(DATA_DIR, 'genome_summary')\n",
    "GENOME_METADATA = path.join(DATA_DIR, 'genome_metadata')\n",
    "\n",
    "SAMPLES_DIR = path.join(DATA_DIR, 'samples')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c49c4402-0ce2-498a-93f4-58d534f7ce2d",
   "metadata": {},
   "source": [
    "## 1. Samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e31a321-bd56-43c8-b6d5-3440e48e04c9",
   "metadata": {},
   "source": [
    "### 1.1 Genomes selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "50e51116-b15d-47ad-a49a-c6abd94d1774",
   "metadata": {},
   "outputs": [],
   "source": [
    "dtypes = {\n",
    "    'genome_id': str, \n",
    "    'genome_name': str,\n",
    "    # 'taxon_id' 'genome_status'\n",
    "    'genome_length': np.float64,\n",
    "    'gc_content': np.float64,\n",
    "    # 'contig_l50' 'contig_n50' 'chromosomes' 'plasmids' 'contigs'\n",
    "    # 'patric_cds' 'refseq_cds' 'trna' 'rrnacoarse_consistency'\n",
    "    'fine_consistency': np.float64,\n",
    "    'checkm_completeness': np.float64,\n",
    "    'checkm_contamination': np.float64,\n",
    "    # 'genome_qualitydate_created' 'date_modified'\n",
    "}\n",
    "\n",
    "df = pd.read_csv(\n",
    "    GENOME_SUMMARY, \n",
    "    sep='\\t',\n",
    "    usecols=dtypes.keys(),\n",
    "    dtype=dtypes\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "00f66a38-3d25-4524-80dd-149f0aaa14c0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>genome_length</th>\n",
       "      <th>gc_content</th>\n",
       "      <th>fine_consistency</th>\n",
       "      <th>checkm_completeness</th>\n",
       "      <th>checkm_contamination</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>315612</th>\n",
       "      <td>1733.752</td>\n",
       "      <td>Mycobacterium tuberculosis  G-032_S_1</td>\n",
       "      <td>4322020.0</td>\n",
       "      <td>65.55000</td>\n",
       "      <td>98.4</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315613</th>\n",
       "      <td>1733.749</td>\n",
       "      <td>Mycobacterium tuberculosis  G-036_S_2</td>\n",
       "      <td>4327952.0</td>\n",
       "      <td>65.55000</td>\n",
       "      <td>98.4</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315614</th>\n",
       "      <td>1733.764</td>\n",
       "      <td>Mycobacterium tuberculosis  G-038_S-1</td>\n",
       "      <td>4328089.0</td>\n",
       "      <td>65.54000</td>\n",
       "      <td>98.3</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315615</th>\n",
       "      <td>1733.797</td>\n",
       "      <td>Mycobacterium tuberculosis  G-039-C</td>\n",
       "      <td>4324832.0</td>\n",
       "      <td>65.54000</td>\n",
       "      <td>98.6</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315616</th>\n",
       "      <td>1733.815</td>\n",
       "      <td>Mycobacterium tuberculosis  G-039-E</td>\n",
       "      <td>4331941.0</td>\n",
       "      <td>65.52000</td>\n",
       "      <td>98.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346926</th>\n",
       "      <td>1773.2434</td>\n",
       "      <td>Mycobacterium tuberculosis strain UM-CSF15</td>\n",
       "      <td>4374121.0</td>\n",
       "      <td>65.61000</td>\n",
       "      <td>99.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346927</th>\n",
       "      <td>1773.2432</td>\n",
       "      <td>Mycobacterium tuberculosis strain UM-CSF17</td>\n",
       "      <td>4356783.0</td>\n",
       "      <td>65.58000</td>\n",
       "      <td>99.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346944</th>\n",
       "      <td>1773.10508</td>\n",
       "      <td>Mycobacterium tuberculosis strain UT1</td>\n",
       "      <td>4357245.0</td>\n",
       "      <td>65.61882</td>\n",
       "      <td>99.4</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346967</th>\n",
       "      <td>1773.213</td>\n",
       "      <td>Mycobacterium tuberculosis strain VRFCWCF XDRT...</td>\n",
       "      <td>4369955.0</td>\n",
       "      <td>65.50000</td>\n",
       "      <td>99.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>346990</th>\n",
       "      <td>1773.834</td>\n",
       "      <td>Mycobacterium tuberculosis strain ZT272</td>\n",
       "      <td>4346549.0</td>\n",
       "      <td>65.54000</td>\n",
       "      <td>98.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5906 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         genome_id                                        genome_name  \\\n",
       "315612    1733.752              Mycobacterium tuberculosis  G-032_S_1   \n",
       "315613    1733.749              Mycobacterium tuberculosis  G-036_S_2   \n",
       "315614    1733.764              Mycobacterium tuberculosis  G-038_S-1   \n",
       "315615    1733.797                Mycobacterium tuberculosis  G-039-C   \n",
       "315616    1733.815                Mycobacterium tuberculosis  G-039-E   \n",
       "...            ...                                                ...   \n",
       "346926   1773.2434         Mycobacterium tuberculosis strain UM-CSF15   \n",
       "346927   1773.2432         Mycobacterium tuberculosis strain UM-CSF17   \n",
       "346944  1773.10508              Mycobacterium tuberculosis strain UT1   \n",
       "346967    1773.213  Mycobacterium tuberculosis strain VRFCWCF XDRT...   \n",
       "346990    1773.834            Mycobacterium tuberculosis strain ZT272   \n",
       "\n",
       "        genome_length  gc_content  fine_consistency  checkm_completeness  \\\n",
       "315612      4322020.0    65.55000              98.4                100.0   \n",
       "315613      4327952.0    65.55000              98.4                100.0   \n",
       "315614      4328089.0    65.54000              98.3                100.0   \n",
       "315615      4324832.0    65.54000              98.6                100.0   \n",
       "315616      4331941.0    65.52000              98.2                100.0   \n",
       "...               ...         ...               ...                  ...   \n",
       "346926      4374121.0    65.61000              99.2                100.0   \n",
       "346927      4356783.0    65.58000              99.0                100.0   \n",
       "346944      4357245.0    65.61882              99.4                100.0   \n",
       "346967      4369955.0    65.50000              99.5                100.0   \n",
       "346990      4346549.0    65.54000              98.5                100.0   \n",
       "\n",
       "        checkm_contamination  \n",
       "315612                   0.0  \n",
       "315613                   0.0  \n",
       "315614                   0.0  \n",
       "315615                   0.0  \n",
       "315616                   0.0  \n",
       "...                      ...  \n",
       "346926                   0.0  \n",
       "346927                   0.0  \n",
       "346944                   0.0  \n",
       "346967                   0.0  \n",
       "346990                   0.0  \n",
       "\n",
       "[5906 rows x 7 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def pipe(df, funcs):\n",
    "    for f in funcs:\n",
    "        df = f(df)\n",
    "    return df\n",
    "\n",
    "\n",
    "def f_mtb(df):\n",
    "    return df[df.genome_name.str.contains('Mycobacterium tuberculosis')]\n",
    "\n",
    "\n",
    "def f_gc_content(df):\n",
    "    # https://bionumbers.hms.harvard.edu/bionumber.aspx?id=101990\n",
    "    value = 65.6\n",
    "    lower_bound = value - 0.1\n",
    "    upper_bound = value + 0.1\n",
    "    \n",
    "    return df[(df.gc_content >= lower_bound) & (df.gc_content <= upper_bound)]\n",
    "\n",
    "\n",
    "def f_genome_length(df):\n",
    "    # https://bionumbers.hms.harvard.edu/bionumber.aspx?id=101741\n",
    "    value = 4.41e6\n",
    "    lower_bound = value - 0.1e6\n",
    "    upper_bound = value + 0.1e6\n",
    "    \n",
    "    return df[(df.genome_length >= lower_bound) & (df.genome_length <= upper_bound)]\n",
    "\n",
    "\n",
    "def f_qc(df):\n",
    "    # https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0250092\n",
    "    # Use more strict QC policy\n",
    "    contamination_threshold = 10 - 9\n",
    "    completeness_threshold = 80 + 19\n",
    "    consistency_threshold = 87 + 11\n",
    "    \n",
    "    df = df[df.checkm_contamination <= contamination_threshold]\n",
    "    df = df[df.checkm_completeness >= completeness_threshold]\n",
    "    df = df[df.fine_consistency >= consistency_threshold]\n",
    "    return df\n",
    "\n",
    "\n",
    "df = pipe(df, [\n",
    "    f_mtb, \n",
    "    f_gc_content, \n",
    "    f_genome_length,\n",
    "    f_qc,\n",
    "])\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e0c7438-dd5f-4bdc-9d29-12c52d6ed271",
   "metadata": {},
   "source": [
    "### 1.2 Genomes metadata enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7f9995e0-e685-4c29-a590-05c1a4d23cfa",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>genome_length</th>\n",
       "      <th>gc_content</th>\n",
       "      <th>fine_consistency</th>\n",
       "      <th>checkm_completeness</th>\n",
       "      <th>checkm_contamination</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>strain</th>\n",
       "      <th>completion_date</th>\n",
       "      <th>...</th>\n",
       "      <th>sequences</th>\n",
       "      <th>patric_cds</th>\n",
       "      <th>refseq_cds</th>\n",
       "      <th>isolation_site</th>\n",
       "      <th>isolation_source</th>\n",
       "      <th>isolation_comments</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>geographic_location</th>\n",
       "      <th>host_name</th>\n",
       "      <th>comments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1733.752</td>\n",
       "      <td>Mycobacterium tuberculosis  G-032_S_1</td>\n",
       "      <td>4322020.0</td>\n",
       "      <td>65.55000</td>\n",
       "      <td>98.4</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1773</td>\n",
       "      <td>G-032_S_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4313.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1733.749</td>\n",
       "      <td>Mycobacterium tuberculosis  G-036_S_2</td>\n",
       "      <td>4327952.0</td>\n",
       "      <td>65.55000</td>\n",
       "      <td>98.4</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1773</td>\n",
       "      <td>G-036_S_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4313.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1733.764</td>\n",
       "      <td>Mycobacterium tuberculosis  G-038_S-1</td>\n",
       "      <td>4328089.0</td>\n",
       "      <td>65.54000</td>\n",
       "      <td>98.3</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1773</td>\n",
       "      <td>G-038_S-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4318.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1733.797</td>\n",
       "      <td>Mycobacterium tuberculosis  G-039-C</td>\n",
       "      <td>4324832.0</td>\n",
       "      <td>65.54000</td>\n",
       "      <td>98.6</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1773</td>\n",
       "      <td>G-039-C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4311.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1733.815</td>\n",
       "      <td>Mycobacterium tuberculosis  G-039-E</td>\n",
       "      <td>4331941.0</td>\n",
       "      <td>65.52000</td>\n",
       "      <td>98.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1773</td>\n",
       "      <td>G-039-E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4318.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5901</th>\n",
       "      <td>1773.2434</td>\n",
       "      <td>Mycobacterium tuberculosis strain UM-CSF15</td>\n",
       "      <td>4374121.0</td>\n",
       "      <td>65.61000</td>\n",
       "      <td>99.2</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1773</td>\n",
       "      <td>UM-CSF15</td>\n",
       "      <td>2016-05-09T00:00:00Z</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4321.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cerebrospinal fluid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human, Homo sapiens</td>\n",
       "      <td>Genomic differences between pulmonary and meni...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5902</th>\n",
       "      <td>1773.2432</td>\n",
       "      <td>Mycobacterium tuberculosis strain UM-CSF17</td>\n",
       "      <td>4356783.0</td>\n",
       "      <td>65.58000</td>\n",
       "      <td>99.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1773</td>\n",
       "      <td>UM-CSF17</td>\n",
       "      <td>2016-05-09T00:00:00Z</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4308.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cerebrospinal fluid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Human, Homo sapiens</td>\n",
       "      <td>Genomic differences between pulmonary and meni...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5903</th>\n",
       "      <td>1773.10508</td>\n",
       "      <td>Mycobacterium tuberculosis strain UT1</td>\n",
       "      <td>4357245.0</td>\n",
       "      <td>65.61882</td>\n",
       "      <td>99.4</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1773</td>\n",
       "      <td>UT1</td>\n",
       "      <td>2018-07-18T00:00:00Z</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4326.0</td>\n",
       "      <td>4063.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Sputum</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2005-06</td>\n",
       "      <td>Colombia:Medellin</td>\n",
       "      <td>Human, Homo sapiens</td>\n",
       "      <td>Determinaci�n de un \"core\" gen�tico de Mycobac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5904</th>\n",
       "      <td>1773.213</td>\n",
       "      <td>Mycobacterium tuberculosis strain VRFCWCF XDRT...</td>\n",
       "      <td>4369955.0</td>\n",
       "      <td>65.50000</td>\n",
       "      <td>99.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1773</td>\n",
       "      <td>VRFCWCF XDRTB 232</td>\n",
       "      <td>2014-06-27T00:00:00Z</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4316.0</td>\n",
       "      <td>4101.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sputum</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2010</td>\n",
       "      <td>India: Chennai</td>\n",
       "      <td>Human, Homo sapiens</td>\n",
       "      <td>THIS STUDY AIMS TO EXPLORE THE DRUG RESISTANT ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5905</th>\n",
       "      <td>1773.834</td>\n",
       "      <td>Mycobacterium tuberculosis strain ZT272</td>\n",
       "      <td>4346549.0</td>\n",
       "      <td>65.54000</td>\n",
       "      <td>98.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1773</td>\n",
       "      <td>ZT272</td>\n",
       "      <td>2015-08-18T00:00:00Z</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4420.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sputum</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2002</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Human, Homo sapiens</td>\n",
       "      <td>Study of M.tuberculosis Beijing</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5906 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       genome_id                                        genome_name  \\\n",
       "0       1733.752              Mycobacterium tuberculosis  G-032_S_1   \n",
       "1       1733.749              Mycobacterium tuberculosis  G-036_S_2   \n",
       "2       1733.764              Mycobacterium tuberculosis  G-038_S-1   \n",
       "3       1733.797                Mycobacterium tuberculosis  G-039-C   \n",
       "4       1733.815                Mycobacterium tuberculosis  G-039-E   \n",
       "...          ...                                                ...   \n",
       "5901   1773.2434         Mycobacterium tuberculosis strain UM-CSF15   \n",
       "5902   1773.2432         Mycobacterium tuberculosis strain UM-CSF17   \n",
       "5903  1773.10508              Mycobacterium tuberculosis strain UT1   \n",
       "5904    1773.213  Mycobacterium tuberculosis strain VRFCWCF XDRT...   \n",
       "5905    1773.834            Mycobacterium tuberculosis strain ZT272   \n",
       "\n",
       "      genome_length  gc_content  fine_consistency  checkm_completeness  \\\n",
       "0         4322020.0    65.55000              98.4                100.0   \n",
       "1         4327952.0    65.55000              98.4                100.0   \n",
       "2         4328089.0    65.54000              98.3                100.0   \n",
       "3         4324832.0    65.54000              98.6                100.0   \n",
       "4         4331941.0    65.52000              98.2                100.0   \n",
       "...             ...         ...               ...                  ...   \n",
       "5901      4374121.0    65.61000              99.2                100.0   \n",
       "5902      4356783.0    65.58000              99.0                100.0   \n",
       "5903      4357245.0    65.61882              99.4                100.0   \n",
       "5904      4369955.0    65.50000              99.5                100.0   \n",
       "5905      4346549.0    65.54000              98.5                100.0   \n",
       "\n",
       "      checkm_contamination taxon_id             strain       completion_date  \\\n",
       "0                      0.0     1773          G-032_S_1                   NaN   \n",
       "1                      0.0     1773          G-036_S_2                   NaN   \n",
       "2                      0.0     1773          G-038_S-1                   NaN   \n",
       "3                      0.0     1773            G-039-C                   NaN   \n",
       "4                      0.0     1773            G-039-E                   NaN   \n",
       "...                    ...      ...                ...                   ...   \n",
       "5901                   0.0     1773           UM-CSF15  2016-05-09T00:00:00Z   \n",
       "5902                   0.0     1773           UM-CSF17  2016-05-09T00:00:00Z   \n",
       "5903                   0.0     1773                UT1  2018-07-18T00:00:00Z   \n",
       "5904                   0.0     1773  VRFCWCF XDRTB 232  2014-06-27T00:00:00Z   \n",
       "5905                   0.0     1773              ZT272  2015-08-18T00:00:00Z   \n",
       "\n",
       "      ... sequences patric_cds refseq_cds isolation_site     isolation_source  \\\n",
       "0     ...       NaN     4313.0        NaN            NaN                  NaN   \n",
       "1     ...       NaN     4313.0        NaN            NaN                  NaN   \n",
       "2     ...       NaN     4318.0        NaN            NaN                  NaN   \n",
       "3     ...       NaN     4311.0        NaN            NaN                  NaN   \n",
       "4     ...       NaN     4318.0        NaN            NaN                  NaN   \n",
       "...   ...       ...        ...        ...            ...                  ...   \n",
       "5901  ...       NaN     4321.0        NaN            NaN  cerebrospinal fluid   \n",
       "5902  ...       NaN     4308.0        NaN            NaN  cerebrospinal fluid   \n",
       "5903  ...       NaN     4326.0     4063.0            NaN               Sputum   \n",
       "5904  ...       NaN     4316.0     4101.0            NaN               sputum   \n",
       "5905  ...       NaN     4420.0        NaN            NaN               sputum   \n",
       "\n",
       "     isolation_comments collection_date geographic_location  \\\n",
       "0                   NaN             NaN                 NaN   \n",
       "1                   NaN             NaN                 NaN   \n",
       "2                   NaN             NaN                 NaN   \n",
       "3                   NaN             NaN                 NaN   \n",
       "4                   NaN             NaN                 NaN   \n",
       "...                 ...             ...                 ...   \n",
       "5901                NaN             NaN                 NaN   \n",
       "5902                NaN             NaN                 NaN   \n",
       "5903                NaN         2005-06   Colombia:Medellin   \n",
       "5904                NaN            2010      India: Chennai   \n",
       "5905                NaN            2002              Brazil   \n",
       "\n",
       "                host_name                                           comments  \n",
       "0                     NaN                                                NaN  \n",
       "1                     NaN                                                NaN  \n",
       "2                     NaN                                                NaN  \n",
       "3                     NaN                                                NaN  \n",
       "4                     NaN                                                NaN  \n",
       "...                   ...                                                ...  \n",
       "5901  Human, Homo sapiens  Genomic differences between pulmonary and meni...  \n",
       "5902  Human, Homo sapiens  Genomic differences between pulmonary and meni...  \n",
       "5903  Human, Homo sapiens  Determinaci�n de un \"core\" gen�tico de Mycobac...  \n",
       "5904  Human, Homo sapiens  THIS STUDY AIMS TO EXPLORE THE DRUG RESISTANT ...  \n",
       "5905  Human, Homo sapiens                    Study of M.tuberculosis Beijing  \n",
       "\n",
       "[5906 rows x 33 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dtypes = {\n",
    "    'genome_id': str,\n",
    "    # 'genome_name' 'organism_name' \n",
    "    'taxon_id': str, \n",
    "    # 'genome_status'\n",
    "    'strain': str,\n",
    "    # 'serovar' 'biovar' 'pathovar' 'mlst' 'other_typing' 'culture_collection' 'type_strain'\n",
    "    'completion_date': str,\n",
    "    'publication': str,\n",
    "    'bioproject_accession': str,\n",
    "    'biosample_accession': str,\n",
    "    'assembly_accession': str,\n",
    "    'genbank_accessions': str,\n",
    "    'refseq_accessions': str,\n",
    "    'sequencing_centers': str,\n",
    "    'sequencing_status': str,\n",
    "    'sequencing_platform': str,\n",
    "    'sequencing_depth': str,\n",
    "    'assembly_method': str,\n",
    "    'chromosomes': str,\n",
    "    # 'plasmids'\n",
    "    'contigs': np.float64,\n",
    "    'sequences': np.float64,\n",
    "    # 'genome_length' 'gc_content'\n",
    "    'patric_cds': np.float64,\n",
    "    # 'brc1_cds'\n",
    "    'refseq_cds': np.float64,\n",
    "    'isolation_site': str,\n",
    "    'isolation_source': str,\n",
    "    'isolation_comments': str,\n",
    "    'collection_date': str,\n",
    "    # 'isolation_country'\n",
    "    'geographic_location': str,\n",
    "    # 'latitude' 'longitude' 'altitude' 'depth' 'other_environmental'\n",
    "    'host_name': str,\n",
    "    # 'host_gender' 'host_age' 'host_health' 'body_sample_site' 'body_sample_subsite' 'other_clinical'\n",
    "    # 'antimicrobial_resistance' 'antimicrobial_resistance_evidence'\n",
    "    # 'gram_stain' 'cell_shape' 'motility' 'sporulation' 'temperature_range' 'optimal_temperature'\n",
    "    # 'salinity' 'oxygen_requirement' 'habitat' 'disease'\n",
    "    'comments': str,\n",
    "    # 'additional_metadata'\n",
    "}\n",
    "\n",
    "df_metadata = pd.read_csv(\n",
    "    GENOME_METADATA, \n",
    "    sep='\\t',\n",
    "    usecols=dtypes.keys(),\n",
    "    dtype=dtypes\n",
    ")\n",
    "\n",
    "df = pd.merge(df, df_metadata, how='inner', on='genome_id')\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0683d3d-cc4d-4dbe-8309-596b19d87182",
   "metadata": {},
   "source": [
    "### 1.3 AMR selection and the most common genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7213aeee-5a6f-49e6-a85c-71ce61c5725b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "dtypes = {\n",
    "    'genome_id': str,\n",
    "    'antibiotic': str,\n",
    "    'resistant_phenotype': str,\n",
    "    'measurement': str,\n",
    "    'measurement_sign': str,\n",
    "    'measurement_unit': str,\n",
    "    'laboratory_typing_method': str,\n",
    "    'laboratory_typing_method_version': str,\n",
    "    'laboratory_typing_platform': str,\n",
    "    'vendor': str,\n",
    "    'testing_standard': str,\n",
    "    'testing_standard_year': str,\n",
    "    'source': str,\n",
    "}\n",
    "\n",
    "df_amr = pd.read_csv(\n",
    "    PATRIC_AMR, \n",
    "    sep='\\t',\n",
    "    usecols=dtypes.keys(),\n",
    "    dtype=dtypes\n",
    ")\n",
    "\n",
    "# Selected genomes\n",
    "df_amr = pd.merge(df[['genome_id', 'genome_name']], df_amr, how='inner', on='genome_id')\n",
    "\n",
    "# Phenotypes with R/S only\n",
    "df_amr = df_amr[df_amr.resistant_phenotype.isin(['Resistant', 'Susceptible'])]\n",
    "\n",
    "# Transform 'laboratory_typing_method_version' in to 'measurement' and 'measurement_unit'\n",
    "convertions = {}\n",
    "for t in df_amr[['measurement', 'measurement_unit', 'laboratory_typing_method_version']].itertuples():\n",
    "    if t.laboratory_typing_method_version is not np.nan:\n",
    "        m = t.laboratory_typing_method_version.split()\n",
    "        # assert len(m) > 1, f\"Not a measurement: {t}\"\n",
    "        if len(m) > 1:\n",
    "            convertions[t.Index] = {'measurement': m[0], 'unit': m[1]}\n",
    "for idx in convertions:\n",
    "    df_amr.loc[idx, 'measurement'] = convertions[idx]['measurement']\n",
    "    df_amr.loc[idx, 'measurement_unit'] = convertions[idx]['unit']\n",
    "    \n",
    "# Turn 'measurement' column dtype into valid number dtype\n",
    "extract_number = lambda text: re.findall(r'[-+]?(?:\\d*\\.*\\d)', text)[0]\n",
    "df_amr['measurement'] = pd.to_numeric(df_amr.measurement.map(extract_number, na_action='ignore'))\n",
    "\n",
    "# Remove sample without measurement\n",
    "df_amr = df_amr.dropna(subset=['measurement'])\n",
    "\n",
    "# Assign NA measurement sign with '='\n",
    "df_amr['measurement_sign'] = df_amr['measurement_sign'].fillna('=')\n",
    "\n",
    "# Use only sample in selected genomes\n",
    "df_amr = pd.merge(df_amr, df['genome_id'], how='inner', on='genome_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "07477f68-2e4e-412b-be5b-e00324bbd3ba",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>resistant</th>\n",
       "      <th>susceptible</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>nicotinamide</th>\n",
       "      <td>77</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ethambutol</th>\n",
       "      <td>280</td>\n",
       "      <td>600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>moxifloxacin</th>\n",
       "      <td>90</td>\n",
       "      <td>197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rifampin</th>\n",
       "      <td>513</td>\n",
       "      <td>401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>capreomycin</th>\n",
       "      <td>86</td>\n",
       "      <td>159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clofazimine</th>\n",
       "      <td>21</td>\n",
       "      <td>262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>streptomycin</th>\n",
       "      <td>332</td>\n",
       "      <td>648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cycloserine</th>\n",
       "      <td>17</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ofloxacin</th>\n",
       "      <td>186</td>\n",
       "      <td>337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ethionamide</th>\n",
       "      <td>188</td>\n",
       "      <td>210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>para-aminosalicylic acid</th>\n",
       "      <td>48</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoniazid</th>\n",
       "      <td>983</td>\n",
       "      <td>622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rifabutin</th>\n",
       "      <td>51</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kanamycin</th>\n",
       "      <td>151</td>\n",
       "      <td>367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ciprofloxacin</th>\n",
       "      <td>17</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>amikacin</th>\n",
       "      <td>44</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          resistant  susceptible\n",
       "nicotinamide                     77           74\n",
       "ethambutol                      280          600\n",
       "moxifloxacin                     90          197\n",
       "rifampin                        513          401\n",
       "capreomycin                      86          159\n",
       "clofazimine                      21          262\n",
       "streptomycin                    332          648\n",
       "cycloserine                      17          109\n",
       "ofloxacin                       186          337\n",
       "ethionamide                     188          210\n",
       "para-aminosalicylic acid         48          109\n",
       "isoniazid                       983          622\n",
       "rifabutin                        51           65\n",
       "kanamycin                       151          367\n",
       "ciprofloxacin                    17           37\n",
       "amikacin                         44          101"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Plot the distributions\n",
    "def phenotype_counts(df_amr):\n",
    "    dist = defaultdict(int)\n",
    "    for d in df_amr[['antibiotic', 'resistant_phenotype']].itertuples():\n",
    "        name = f'{d.antibiotic}:{d.resistant_phenotype.lower()}'\n",
    "        dist[name] += 1\n",
    "    \n",
    "    antibiotics = set([a.split(':')[0] for a in dist.keys()])\n",
    "    phenotype_counts = [(dist[f'{a}:resistant'], dist[f'{a}:susceptible']) for a in antibiotics]\n",
    "    return pd.DataFrame(phenotype_counts, index=list(antibiotics), columns=['resistant', 'susceptible'])\n",
    "\n",
    "\n",
    "phenotype_counts(df_amr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "67761b29-9973-4c66-b4bc-578d3a9ae7b0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "selected_antibiotics = ['isoniazid', 'streptomycin', 'rifampin', 'ethambutol']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "58631b69-495e-4f5f-a3fe-26e88454cc90",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "710"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Presences of the most common genomes in major antibiotics\n",
    "def get_genomes(df_amr, antibiotic):\n",
    "    return df_amr[df_amr.antibiotic.str.contains(antibiotic)].genome_id\n",
    "\n",
    "g_isoniazid = get_genomes(df_amr, 'isoniazid')\n",
    "assert len(g_isoniazid) == 983 + 622\n",
    "g_streptomycin = get_genomes(df_amr, 'streptomycin')\n",
    "assert len(g_streptomycin) == 332 + 648\n",
    "g_rifampin = get_genomes(df_amr, 'rifampin')\n",
    "assert len(g_rifampin) == 513 + 401\n",
    "g_ethambutol = get_genomes(df_amr, 'ethambutol')\n",
    "assert len(g_ethambutol) == 280 + 600\n",
    "\n",
    "common_genomes = set(g_isoniazid)\n",
    "common_genomes.intersection_update(g_streptomycin)\n",
    "common_genomes.intersection_update(g_rifampin)\n",
    "common_genomes.intersection_update(g_ethambutol)\n",
    "\n",
    "len(common_genomes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ed33c735-bf39-41e2-80dc-9747ce962a47",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>resistant</th>\n",
       "      <th>susceptible</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>streptomycin</th>\n",
       "      <td>271</td>\n",
       "      <td>616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoniazid</th>\n",
       "      <td>634</td>\n",
       "      <td>589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ethambutol</th>\n",
       "      <td>277</td>\n",
       "      <td>590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rifampin</th>\n",
       "      <td>339</td>\n",
       "      <td>381</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              resistant  susceptible\n",
       "streptomycin        271          616\n",
       "isoniazid           634          589\n",
       "ethambutol          277          590\n",
       "rifampin            339          381"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_amr = df_amr[df_amr.genome_id.isin(common_genomes) & df_amr.antibiotic.isin(selected_antibiotics)]\n",
    "\n",
    "phenotype_counts(df_amr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3385a1f1-68be-4d34-acae-69f41b6670ff",
   "metadata": {},
   "source": [
    "### 1.4 Clinical breakpoints"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "151764ae-f995-4f26-9a3e-736728831320",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Agar proportion' 'Agar dilution' 'Broth dilution']\n",
      "\n",
      "['ethambutol' 'isoniazid' 'rifampin' 'streptomycin']\n",
      "\n",
      "['Resistant' 'Susceptible']\n"
     ]
    }
   ],
   "source": [
    "print(\n",
    "    df_amr.laboratory_typing_method.unique(),\n",
    "    df_amr.antibiotic.unique(),\n",
    "    df_amr.resistant_phenotype.unique(),\n",
    "    sep='\\n\\n'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "084c5eea-637d-4d42-9f60-9787427dfe7d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Breakpoints/Critical concentrations tables\n",
    "# https://www.sciencedirect.com/science/article/pii/S1198743X16305110#tbl2\n",
    "# https://www.idexx.com/media/filer_public/b2/04/b2047e39-bbf3-48c1-8c26-076b7f850922/microbiology-guide-interpreting-mic.pdf\n",
    "# \n",
    "# https://en.wikipedia.org/wiki/Middlebrook_7H9_Broth\n",
    "# https://en.wikipedia.org/wiki/Middlebrook_7H10_Agar\n",
    "breakpoints = {\n",
    "    'ethambutol': {'7H10': 5.0, '7H9': 5.0},\n",
    "    'isoniazid': {'7H10': 0.2, '7H9': 0.1},\n",
    "    'rifampin': {'7H10': 1.0,'7H9': 1.0},\n",
    "    'streptomycin': {'7H10': 2.0, '7H9': 1.0},\n",
    "}\n",
    "\n",
    "# Rules:\n",
    "#   Isolates that cannot grow at ‘critical’ concentrations were then defined as susceptible, \n",
    "#   whereas those that can grow were considered resistant\n",
    "# \n",
    "# S = Susceptible\n",
    "# R = Resistant\n",
    "# NR = Non-resistant when the phenotype is S but the its MIC is NOT < breakpoint\n",
    "# NS = Non-susceptible when the phenotype is R but the its MIC is NOT >= breakpoint*2\n",
    "def reannotate_phenotype(measurement, antibiotic, phenotype, method):\n",
    "    def validate_by_beakpoint(antibiotic, media):\n",
    "        if phenotype == 'Susceptible':\n",
    "            if not measurement <= breakpoints[antibiotic][media]: \n",
    "                return 'NR'\n",
    "            else:\n",
    "                return 'S'\n",
    "        if phenotype == 'Resistant':\n",
    "            if not measurement >= 2*breakpoints[antibiotic][media]:\n",
    "                return 'NS'\n",
    "            else:\n",
    "                return 'R'\n",
    "    \n",
    "    if method == 'Agar proportion' or method == 'Agar dilution':\n",
    "        return validate_by_beakpoint(antibiotic, '7H10')\n",
    "    if method == 'Broth dilution':\n",
    "        return validate_by_beakpoint(antibiotic, '7H9')\n",
    "    \n",
    "    \n",
    "assert reannotate_phenotype(2.0, 'ethambutol', 'Resistant', 'Agar proportion') == 'NS' \n",
    "assert reannotate_phenotype(4.0, 'streptomycin', 'Susceptible', 'Agar dilution') == 'NR' \n",
    "assert reannotate_phenotype(1.0, 'rifampin', 'Resistant', 'Agar dilution') == 'NS'\n",
    "assert reannotate_phenotype(2.0, 'rifampin', 'Resistant', 'Agar dilution') == 'R'\n",
    "assert reannotate_phenotype(1.0, 'rifampin', 'Susceptible', 'Agar dilution') == 'S'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "2ce03479-085d-45c6-95e0-490bba4c6fb9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "phenotype\n",
       "NR    1299\n",
       "NS    1054\n",
       "S      877\n",
       "R      467\n",
       "dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_amr['phenotype'] = df_amr.apply(\n",
    "    lambda r: reannotate_phenotype(\n",
    "        r['measurement'], \n",
    "        r['antibiotic'], \n",
    "        r['resistant_phenotype'], \n",
    "        r['laboratory_typing_method']\n",
    "    ),\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "df_amr.value_counts('phenotype')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2740f020-b772-4467-a6e4-76731d4798b6",
   "metadata": {},
   "source": [
    "### 1.6 Only use valid samples (without NR and NS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "346392ca-705a-45b0-844e-b37027718a82",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genomes: 510\n",
      "Valid tests: 1344\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>resistant</th>\n",
       "      <th>susceptible</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>streptomycin</th>\n",
       "      <td>89</td>\n",
       "      <td>288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoniazid</th>\n",
       "      <td>251</td>\n",
       "      <td>270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ethambutol</th>\n",
       "      <td>41</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rifampin</th>\n",
       "      <td>86</td>\n",
       "      <td>207</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              resistant  susceptible\n",
       "streptomycin         89          288\n",
       "isoniazid           251          270\n",
       "ethambutol           41          112\n",
       "rifampin             86          207"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_amr_valid = df_amr[~df_amr.phenotype.isin(['NS', 'NR'])]\n",
    "used_genomes = df_amr_valid.genome_id.unique()\n",
    "\n",
    "print(f'Genomes: {len(used_genomes)}')\n",
    "print(f'Valid tests: {len(df_amr_valid)}')\n",
    "phenotype_counts(df_amr_valid)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbb0b93a-d794-42e3-a4a9-0acfb0462b77",
   "metadata": {},
   "source": [
    "### 1.7 Remove extreme inconsistence samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "722aa6d5-e28a-4b1c-81c8-44de9497ddf6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genomes: 510\n",
      "Valid tests: 1243\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>resistant</th>\n",
       "      <th>susceptible</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>streptomycin</th>\n",
       "      <td>89</td>\n",
       "      <td>277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoniazid</th>\n",
       "      <td>251</td>\n",
       "      <td>184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ethambutol</th>\n",
       "      <td>41</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rifampin</th>\n",
       "      <td>86</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              resistant  susceptible\n",
       "streptomycin         89          277\n",
       "isoniazid           251          184\n",
       "ethambutol           41          112\n",
       "rifampin             86          203"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_amr_valid = df_amr_valid.drop_duplicates(subset=['genome_id', 'antibiotic', 'phenotype'])\n",
    "\n",
    "print(f'Genomes: {len(df_amr_valid.genome_id.unique())}')\n",
    "print(f'Valid tests: {len(df_amr_valid)}')\n",
    "phenotype_counts(df_amr_valid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "61eab114-dc27-4e08-b3e7-1fcf8e88015a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Inconsistency in rifampin:\n",
      "________________________________________________\n",
      "\n",
      "Inconsistency in streptomycin:\n",
      "     genome_id    antibiotic phenotype  measurement laboratory_typing_method\n",
      "7351  1773.203    ethambutol         S          0.1           Broth dilution\n",
      "7355  1773.203     isoniazid         S          0.1           Broth dilution\n",
      "7359  1773.203      rifampin         S          0.1           Broth dilution\n",
      "7361  1773.203  streptomycin         S          0.1           Broth dilution\n",
      "7363  1773.203  streptomycin         R          4.0            Agar dilution\n",
      "________________________________________________\n",
      "\n",
      "Inconsistency in ethambutol:\n",
      "________________________________________________\n",
      "\n",
      "Inconsistency in isoniazid:\n",
      "      genome_id antibiotic phenotype  measurement laboratory_typing_method\n",
      "3160  1324218.3  isoniazid         S          0.2            Agar dilution\n",
      "3162  1324218.3  isoniazid         R          1.0            Agar dilution\n",
      "      genome_id    antibiotic phenotype  measurement laboratory_typing_method\n",
      "3393  1324235.3     isoniazid         S          0.2            Agar dilution\n",
      "3395  1324235.3     isoniazid         R          1.0            Agar dilution\n",
      "3400  1324235.3  streptomycin         S          2.0            Agar dilution\n",
      "________________________________________________\n"
     ]
    }
   ],
   "source": [
    "def report_inconsistency(df_amr_valid):\n",
    "    inconsistence_samples = []\n",
    "\n",
    "    for antibiotic in df_amr_valid.antibiotic.unique():\n",
    "        checker = defaultdict(list)\n",
    "    \n",
    "        for r in df_amr_valid[df_amr_valid.antibiotic == antibiotic].itertuples():\n",
    "            checker[r.genome_id].append(r.phenotype)\n",
    "    \n",
    "        print(f'\\nInconsistency in {antibiotic}:')\n",
    "        for gid in checker.keys():\n",
    "            if len(checker[gid]) > 1:\n",
    "                columns = ['genome_id', 'antibiotic', 'phenotype', 'measurement', 'laboratory_typing_method']\n",
    "                invalid_sample = df_amr_valid.loc[(df_amr.genome_id == gid), columns]\n",
    "                print(invalid_sample)\n",
    "                inconsistence_samples.append(gid)\n",
    "        print('________________________________________________')\n",
    "            \n",
    "            \n",
    "report_inconsistency(df_amr_valid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "bd31c550-a71c-48f7-997c-978df60681d2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Inconsistency in rifampin:\n",
      "________________________________________________\n",
      "\n",
      "Inconsistency in streptomycin:\n",
      "________________________________________________\n",
      "\n",
      "Inconsistency in ethambutol:\n",
      "________________________________________________\n",
      "\n",
      "Inconsistency in isoniazid:\n",
      "________________________________________________\n",
      "Genomes: 507\n",
      "Valid tests: 1233\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>resistant</th>\n",
       "      <th>susceptible</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>streptomycin</th>\n",
       "      <td>88</td>\n",
       "      <td>275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoniazid</th>\n",
       "      <td>249</td>\n",
       "      <td>181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ethambutol</th>\n",
       "      <td>41</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rifampin</th>\n",
       "      <td>86</td>\n",
       "      <td>202</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              resistant  susceptible\n",
       "streptomycin         88          275\n",
       "isoniazid           249          181\n",
       "ethambutol           41          111\n",
       "rifampin             86          202"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bad_samples = ['1773.203', '1324218.3', '1324235.3']\n",
    "\n",
    "df_amr_valid = df_amr_valid[~df_amr_valid.genome_id.isin(bad_samples)]\n",
    "\n",
    "report_inconsistency(df_amr_valid)\n",
    "\n",
    "print(f'Genomes: {len(df_amr_valid.genome_id.unique())}')\n",
    "print(f'Valid tests: {len(df_amr_valid)}')\n",
    "phenotype_counts(df_amr_valid)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20ccce63-793f-4c47-b9b0-217290cf9028",
   "metadata": {},
   "source": [
    "## 2. Results for QC (Mash)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5b81026-93cd-4fe7-840f-4ece2e623639",
   "metadata": {},
   "source": [
    "**REMEBER** to add the genome reference!\n",
    "- Genome ID: **83332.12**\n",
    "- Genome Name: **Mycobacterium tuberculosis H37Rv**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "338f4d9c-bc9e-4cd3-b727-c006af59e1c6",
   "metadata": {},
   "source": [
    "### 2.1 Export to list and also embed the reference genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a4bf03cc-6fc2-4f19-95d9-fdb37bbddb55",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "reference_id = '83332.12'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0fc23063-aee3-4bce-bb7a-5479bf1eea6c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "FNA_LIST_TXT = path.join(SAMPLES_DIR, '_fna_before_mash.txt')\n",
    "GFF_LIST_TXT = path.join(SAMPLES_DIR, '_gff_before_mash.txt')\n",
    "\n",
    "f_fna = open(FNA_LIST_TXT, 'w')\n",
    "f_gff = open(GFF_LIST_TXT, 'w')\n",
    "\n",
    "genome_ids = [reference_id]\n",
    "genome_ids.extend(used_genomes)\n",
    "\n",
    "for gid in genome_ids:\n",
    "    f_fna.write(f'{gid}.fna\\n')\n",
    "    f_gff.write(f'{gid}.PATRIC.gff\\n')\n",
    "\n",
    "f_fna.close()\n",
    "f_gff.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a241903-d136-4a44-bb8a-56d59b6bd660",
   "metadata": {},
   "source": [
    "### 2.2 Export to Excel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b662209f-72fb-4bb8-93e0-2cf5a5512f21",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Export AMR to Excel\n",
    "wanted_columns = [\n",
    "    'genome_id',\n",
    "    'genome_name',\n",
    "    'antibiotic',\n",
    "    'phenotype',\n",
    "    'measurement',\n",
    "    'measurement_unit',\n",
    "    'laboratory_typing_method',\n",
    "]\n",
    "AMR_XLSX = path.join(DATA_DIR, 'amr_before_mash.xlsx')\n",
    "df_amr_valid[wanted_columns].to_excel(AMR_XLSX, index=False)\n",
    "\n",
    "# Export Genomes to Excel\n",
    "wanted_columns = ['genome_id',\n",
    "    'genome_name',\n",
    "    'strain',\n",
    "    'taxon_id', \n",
    "    'genome_length',\n",
    "    'gc_content',\n",
    "    'fine_consistency', \n",
    "    'checkm_completeness', \n",
    "    'checkm_contamination',\n",
    "    'completion_date',\n",
    "    'publication',\n",
    "    'bioproject_accession',\n",
    "    'biosample_accession',\n",
    "    'assembly_accession',\n",
    "    'genbank_accessions',\n",
    "    'refseq_accessions',\n",
    "    'sequencing_centers',\n",
    "    'sequencing_status',\n",
    "    'sequencing_platform',\n",
    "    'sequencing_depth',\n",
    "    'assembly_method',\n",
    "    'chromosomes',\n",
    "    'contigs',\n",
    "    'sequences',\n",
    "    'patric_cds',\n",
    "    'refseq_cds',\n",
    "    'isolation_site',\n",
    "    'isolation_source',\n",
    "    'isolation_comments',\n",
    "    'collection_date',\n",
    "    'geographic_location',\n",
    "    'host_name',\n",
    "    'comments'\n",
    "]\n",
    "GENOMES_XLSX = path.join(DATA_DIR, 'genomes_before_mash.xlsx')\n",
    "df = df[df.genome_id.isin(used_genomes)]\n",
    "df[wanted_columns].to_excel(GENOMES_XLSX, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e368eb5a-37ae-47f6-9ed2-434b37bdb7ec",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
