{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "48c77479-e823-4188-9b54-fa69dda20101",
   "metadata": {},
   "source": [
    "# Panaroo Preparasi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5a23789d-6a92-44df-998e-fedf0ebb2bd2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/hnasrulloh/Works/pangenomes/notebooks'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b6987d04-ce02-4cf8-848a-05e7bbf23a91",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from os import path\n",
    "from collections import defaultdict\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "acfb6c2a-a450-42f6-9bf6-af5af320d2d3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "DATA_DIR = path.join(path.pardir, 'data')\n",
    "\n",
    "SAMPLES_DIR = path.join(DATA_DIR, 'samples')\n",
    "QC_DIR = path.join(DATA_DIR, 'qc')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce5cc5c3-94f4-4b04-a685-585d128914f0",
   "metadata": {},
   "source": [
    "## 1. Deteksi Outlier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f66506cf-4e58-44be-8315-f7f86c91d719",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>total_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1267355.3</td>\n",
       "      <td>4365.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1267356.3</td>\n",
       "      <td>4345.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1267358.3</td>\n",
       "      <td>4358.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1267359.3</td>\n",
       "      <td>4331.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1267364.3</td>\n",
       "      <td>4355.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>707</th>\n",
       "      <td>1773.203</td>\n",
       "      <td>4312.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>708</th>\n",
       "      <td>1773.204</td>\n",
       "      <td>4269.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>709</th>\n",
       "      <td>652616.4</td>\n",
       "      <td>4372.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>710</th>\n",
       "      <td>83331.31</td>\n",
       "      <td>4346.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>711</th>\n",
       "      <td>83332.12</td>\n",
       "      <td>4367.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>712 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     genome_id  total_genes\n",
       "0    1267355.3       4365.0\n",
       "1    1267356.3       4345.0\n",
       "2    1267358.3       4358.0\n",
       "3    1267359.3       4331.0\n",
       "4    1267364.3       4355.0\n",
       "..         ...          ...\n",
       "707   1773.203       4312.0\n",
       "708   1773.204       4269.0\n",
       "709   652616.4       4372.0\n",
       "710   83331.31       4346.0\n",
       "711   83332.12       4367.0\n",
       "\n",
       "[712 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NGENES_TXT = path.join(QC_DIR, 'ngenes.txt')\n",
    "\n",
    "genes = pd.read_csv(NGENES_TXT, sep='\\t', dtype={'sample': str, 'no_genes': np.float64})\n",
    "genes.columns = ['genome_id', 'total_genes']\n",
    "genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5099776d-c1b8-460f-b8d5-13fa34ec9bfe",
   "metadata": {},
   "source": [
    "### 1.1 Outliers Total Gen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4f52e5f5-4601-4206-98ba-c18a32f4d0cc",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>total_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1267355.3</td>\n",
       "      <td>4365.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1267356.3</td>\n",
       "      <td>4345.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1267358.3</td>\n",
       "      <td>4358.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1267359.3</td>\n",
       "      <td>4331.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1267364.3</td>\n",
       "      <td>4355.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>706</th>\n",
       "      <td>1773.202</td>\n",
       "      <td>4307.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>707</th>\n",
       "      <td>1773.203</td>\n",
       "      <td>4312.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>709</th>\n",
       "      <td>652616.4</td>\n",
       "      <td>4372.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>710</th>\n",
       "      <td>83331.31</td>\n",
       "      <td>4346.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>711</th>\n",
       "      <td>83332.12</td>\n",
       "      <td>4367.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>671 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     genome_id  total_genes\n",
       "0    1267355.3       4365.0\n",
       "1    1267356.3       4345.0\n",
       "2    1267358.3       4358.0\n",
       "3    1267359.3       4331.0\n",
       "4    1267364.3       4355.0\n",
       "..         ...          ...\n",
       "706   1773.202       4307.0\n",
       "707   1773.203       4312.0\n",
       "709   652616.4       4372.0\n",
       "710   83331.31       4346.0\n",
       "711   83332.12       4367.0\n",
       "\n",
       "[671 rows x 2 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "upper_outlier = genes[genes['genome_id'] == '1400918.3'].total_genes.to_list()[0]\n",
    "lower_outlier = genes[genes['genome_id'] == '1773.200'].total_genes.to_list()[0]\n",
    "\n",
    "genes = genes[(genes.total_genes <= upper_outlier) & (genes.total_genes >= lower_outlier)]\n",
    "genes"
   ]
  },
  {
   "cell_type": "raw",
   "id": "3cb5f81d-59aa-4749-b330-505efb203ec3",
   "metadata": {},
   "source": [
    "upper: 1400918.3     lower: 1773.200   bad: 1324246.3 1295760.3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef8920d7-4bc7-43d0-a4b7-cce9dcb576ad",
   "metadata": {},
   "source": [
    "### 1.2 Outlier dengan MDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "630975f6-1994-497c-993b-0b91ab25c873",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>total_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1267355.3</td>\n",
       "      <td>4365.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1267356.3</td>\n",
       "      <td>4345.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1267358.3</td>\n",
       "      <td>4358.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1267359.3</td>\n",
       "      <td>4331.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1267364.3</td>\n",
       "      <td>4355.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>706</th>\n",
       "      <td>1773.202</td>\n",
       "      <td>4307.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>707</th>\n",
       "      <td>1773.203</td>\n",
       "      <td>4312.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>709</th>\n",
       "      <td>652616.4</td>\n",
       "      <td>4372.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>710</th>\n",
       "      <td>83331.31</td>\n",
       "      <td>4346.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>711</th>\n",
       "      <td>83332.12</td>\n",
       "      <td>4367.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>669 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     genome_id  total_genes\n",
       "0    1267355.3       4365.0\n",
       "1    1267356.3       4345.0\n",
       "2    1267358.3       4358.0\n",
       "3    1267359.3       4331.0\n",
       "4    1267364.3       4355.0\n",
       "..         ...          ...\n",
       "706   1773.202       4307.0\n",
       "707   1773.203       4312.0\n",
       "709   652616.4       4372.0\n",
       "710   83331.31       4346.0\n",
       "711   83332.12       4367.0\n",
       "\n",
       "[669 rows x 2 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mds_outliers = ['1324246.3', '1295760.3']\n",
    "\n",
    "genes = genes[~genes.genome_id.isin(mds_outliers)]\n",
    "genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f59ba46-c199-473b-b70d-51043ef07a2d",
   "metadata": {},
   "source": [
    "## 2. Export"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2baa075c-7693-4825-b346-67328df55afd",
   "metadata": {},
   "source": [
    "### 2.1 Excel: AMR dan Genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "24e21321-2309-4381-a423-3fea66207bcf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>antibiotic</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>measurement</th>\n",
       "      <th>measurement_unit</th>\n",
       "      <th>laboratory_typing_method</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1438833.3</td>\n",
       "      <td>Mycobacterium tuberculosis 1010SM</td>\n",
       "      <td>rifampin</td>\n",
       "      <td>R</td>\n",
       "      <td>40</td>\n",
       "      <td>mg/L</td>\n",
       "      <td>Agar proportion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1438835.3</td>\n",
       "      <td>Mycobacterium tuberculosis 1173CS</td>\n",
       "      <td>ethambutol</td>\n",
       "      <td>S</td>\n",
       "      <td>2</td>\n",
       "      <td>mg/L</td>\n",
       "      <td>Agar proportion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1438835.3</td>\n",
       "      <td>Mycobacterium tuberculosis 1173CS</td>\n",
       "      <td>rifampin</td>\n",
       "      <td>R</td>\n",
       "      <td>40</td>\n",
       "      <td>mg/L</td>\n",
       "      <td>Agar proportion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1438837.3</td>\n",
       "      <td>Mycobacterium tuberculosis 1429BH</td>\n",
       "      <td>ethambutol</td>\n",
       "      <td>S</td>\n",
       "      <td>2</td>\n",
       "      <td>mg/L</td>\n",
       "      <td>Agar proportion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1438837.3</td>\n",
       "      <td>Mycobacterium tuberculosis 1429BH</td>\n",
       "      <td>rifampin</td>\n",
       "      <td>R</td>\n",
       "      <td>40</td>\n",
       "      <td>mg/L</td>\n",
       "      <td>Agar proportion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1294</th>\n",
       "      <td>1773.200</td>\n",
       "      <td>Mycobacterium tuberculosis strain MAL010077</td>\n",
       "      <td>isoniazid</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>mg/L</td>\n",
       "      <td>Agar dilution</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1298</th>\n",
       "      <td>1773.202</td>\n",
       "      <td>Mycobacterium tuberculosis strain MAL020139</td>\n",
       "      <td>isoniazid</td>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>mg/L</td>\n",
       "      <td>Agar dilution</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1299</th>\n",
       "      <td>1773.203</td>\n",
       "      <td>Mycobacterium tuberculosis strain MAL020151</td>\n",
       "      <td>ethambutol</td>\n",
       "      <td>S</td>\n",
       "      <td>0.1</td>\n",
       "      <td>mg/L</td>\n",
       "      <td>Broth dilution</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1300</th>\n",
       "      <td>1773.203</td>\n",
       "      <td>Mycobacterium tuberculosis strain MAL020151</td>\n",
       "      <td>isoniazid</td>\n",
       "      <td>S</td>\n",
       "      <td>0.1</td>\n",
       "      <td>mg/L</td>\n",
       "      <td>Broth dilution</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1301</th>\n",
       "      <td>1773.203</td>\n",
       "      <td>Mycobacterium tuberculosis strain MAL020151</td>\n",
       "      <td>rifampin</td>\n",
       "      <td>S</td>\n",
       "      <td>0.1</td>\n",
       "      <td>mg/L</td>\n",
       "      <td>Broth dilution</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1210 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      genome_id                                  genome_name  antibiotic  \\\n",
       "0     1438833.3            Mycobacterium tuberculosis 1010SM    rifampin   \n",
       "1     1438835.3            Mycobacterium tuberculosis 1173CS  ethambutol   \n",
       "2     1438835.3            Mycobacterium tuberculosis 1173CS    rifampin   \n",
       "3     1438837.3            Mycobacterium tuberculosis 1429BH  ethambutol   \n",
       "4     1438837.3            Mycobacterium tuberculosis 1429BH    rifampin   \n",
       "...         ...                                          ...         ...   \n",
       "1294   1773.200  Mycobacterium tuberculosis strain MAL010077   isoniazid   \n",
       "1298   1773.202  Mycobacterium tuberculosis strain MAL020139   isoniazid   \n",
       "1299   1773.203  Mycobacterium tuberculosis strain MAL020151  ethambutol   \n",
       "1300   1773.203  Mycobacterium tuberculosis strain MAL020151   isoniazid   \n",
       "1301   1773.203  Mycobacterium tuberculosis strain MAL020151    rifampin   \n",
       "\n",
       "     phenotype measurement measurement_unit laboratory_typing_method  \n",
       "0            R          40             mg/L          Agar proportion  \n",
       "1            S           2             mg/L          Agar proportion  \n",
       "2            R          40             mg/L          Agar proportion  \n",
       "3            S           2             mg/L          Agar proportion  \n",
       "4            R          40             mg/L          Agar proportion  \n",
       "...        ...         ...              ...                      ...  \n",
       "1294         R           1             mg/L            Agar dilution  \n",
       "1298         R           1             mg/L            Agar dilution  \n",
       "1299         S         0.1             mg/L           Broth dilution  \n",
       "1300         S         0.1             mg/L           Broth dilution  \n",
       "1301         S         0.1             mg/L           Broth dilution  \n",
       "\n",
       "[1210 rows x 7 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AMR_BEFORE_MASH = path.join(DATA_DIR, 'amr_before_mash.xlsx')\n",
    "df_amr = pd.read_excel(AMR_BEFORE_MASH, dtype=str)\n",
    "\n",
    "df_amr = df_amr[df_amr.genome_id.isin(genes.genome_id)]\n",
    "\n",
    "AMR_READY = path.join(DATA_DIR, 'amr_ready.xlsx')\n",
    "df_amr.to_excel(AMR_READY, index=False)\n",
    "\n",
    "df_amr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "69b33e71-4090-4224-828a-95847a4bdca4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>strain</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>genome_length</th>\n",
       "      <th>gc_content</th>\n",
       "      <th>fine_consistency</th>\n",
       "      <th>checkm_completeness</th>\n",
       "      <th>checkm_contamination</th>\n",
       "      <th>completion_date</th>\n",
       "      <th>...</th>\n",
       "      <th>sequences</th>\n",
       "      <th>patric_cds</th>\n",
       "      <th>refseq_cds</th>\n",
       "      <th>isolation_site</th>\n",
       "      <th>isolation_source</th>\n",
       "      <th>isolation_comments</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>geographic_location</th>\n",
       "      <th>host_name</th>\n",
       "      <th>comments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1438833.3</td>\n",
       "      <td>Mycobacterium tuberculosis 1010SM</td>\n",
       "      <td>1010SM</td>\n",
       "      <td>1438833</td>\n",
       "      <td>4406079</td>\n",
       "      <td>65.6</td>\n",
       "      <td>99.3</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>2014-04-29T00:00:00Z</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4372</td>\n",
       "      <td>4006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2013</td>\n",
       "      <td>Romania</td>\n",
       "      <td>Human, Homo sapiens</td>\n",
       "      <td>TB-ARC - Romania</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1438835.3</td>\n",
       "      <td>Mycobacterium tuberculosis 1173CS</td>\n",
       "      <td>1173CS</td>\n",
       "      <td>1438835</td>\n",
       "      <td>4383014</td>\n",
       "      <td>65.6</td>\n",
       "      <td>99.2</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>2014-04-29T00:00:00Z</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4333</td>\n",
       "      <td>3982</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bodily fluid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2012</td>\n",
       "      <td>Romania</td>\n",
       "      <td>Human, Homo sapiens</td>\n",
       "      <td>TB-ARC - Romania</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1438837.3</td>\n",
       "      <td>Mycobacterium tuberculosis 1429BH</td>\n",
       "      <td>1429BH</td>\n",
       "      <td>1438837</td>\n",
       "      <td>4417210</td>\n",
       "      <td>65.6</td>\n",
       "      <td>99.3</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>2014-04-29T00:00:00Z</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4384</td>\n",
       "      <td>4022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bodily fluid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2013</td>\n",
       "      <td>Romania</td>\n",
       "      <td>Human, Homo sapiens</td>\n",
       "      <td>TB-ARC - Romania</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1438838.3</td>\n",
       "      <td>Mycobacterium tuberculosis 1430BH</td>\n",
       "      <td>1430BH</td>\n",
       "      <td>1438838</td>\n",
       "      <td>4418670</td>\n",
       "      <td>65.6</td>\n",
       "      <td>99.3</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>2014-04-29T00:00:00Z</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4383</td>\n",
       "      <td>4023</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bodily fluid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2013</td>\n",
       "      <td>Romania</td>\n",
       "      <td>Human, Homo sapiens</td>\n",
       "      <td>TB-ARC - Romania</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1438842.3</td>\n",
       "      <td>Mycobacterium tuberculosis 1755BN</td>\n",
       "      <td>1755BN</td>\n",
       "      <td>1438842</td>\n",
       "      <td>4394400</td>\n",
       "      <td>65.6</td>\n",
       "      <td>99.3</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>2014-04-29T00:00:00Z</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4360</td>\n",
       "      <td>3987</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bodily fluid</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2013</td>\n",
       "      <td>Romania</td>\n",
       "      <td>Human, Homo sapiens</td>\n",
       "      <td>TB-ARC - Romania</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>652616.4</td>\n",
       "      <td>Mycobacterium tuberculosis str. Erdman = ATCC ...</td>\n",
       "      <td>Erdman</td>\n",
       "      <td>652616</td>\n",
       "      <td>4417384</td>\n",
       "      <td>65.65</td>\n",
       "      <td>99.3</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4372</td>\n",
       "      <td>4006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705</th>\n",
       "      <td>1773.199</td>\n",
       "      <td>Mycobacterium tuberculosis strain MAL010075</td>\n",
       "      <td>MAL010075</td>\n",
       "      <td>1773</td>\n",
       "      <td>4421633</td>\n",
       "      <td>65.6</td>\n",
       "      <td>99.3</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>2014-04-29T00:00:00Z</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4320</td>\n",
       "      <td>4006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2007</td>\n",
       "      <td>Mali</td>\n",
       "      <td>Human, Homo sapiens</td>\n",
       "      <td>TB-ARC - Mali</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>706</th>\n",
       "      <td>1773.200</td>\n",
       "      <td>Mycobacterium tuberculosis strain MAL010077</td>\n",
       "      <td>MAL010077</td>\n",
       "      <td>1773</td>\n",
       "      <td>4421308</td>\n",
       "      <td>65.6</td>\n",
       "      <td>99.3</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>2014-04-29T00:00:00Z</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4304</td>\n",
       "      <td>4006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mali</td>\n",
       "      <td>Human, Homo sapiens</td>\n",
       "      <td>TB-ARC - Mali</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>708</th>\n",
       "      <td>1773.202</td>\n",
       "      <td>Mycobacterium tuberculosis strain MAL020139</td>\n",
       "      <td>MAL020139</td>\n",
       "      <td>1773</td>\n",
       "      <td>4413724</td>\n",
       "      <td>65.6</td>\n",
       "      <td>99.4</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>2014-04-29T00:00:00Z</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4307</td>\n",
       "      <td>4003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2007</td>\n",
       "      <td>Mali</td>\n",
       "      <td>Human, Homo sapiens</td>\n",
       "      <td>TB-ARC - Mali</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>709</th>\n",
       "      <td>1773.203</td>\n",
       "      <td>Mycobacterium tuberculosis strain MAL020151</td>\n",
       "      <td>MAL020151</td>\n",
       "      <td>1773</td>\n",
       "      <td>4396046</td>\n",
       "      <td>65.6</td>\n",
       "      <td>99.3</td>\n",
       "      <td>100</td>\n",
       "      <td>0</td>\n",
       "      <td>2014-04-29T00:00:00Z</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4312</td>\n",
       "      <td>4019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2007</td>\n",
       "      <td>Mali</td>\n",
       "      <td>Human, Homo sapiens</td>\n",
       "      <td>TB-ARC - Mali</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>668 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     genome_id                                        genome_name     strain  \\\n",
       "0    1438833.3                  Mycobacterium tuberculosis 1010SM     1010SM   \n",
       "1    1438835.3                  Mycobacterium tuberculosis 1173CS     1173CS   \n",
       "2    1438837.3                  Mycobacterium tuberculosis 1429BH     1429BH   \n",
       "3    1438838.3                  Mycobacterium tuberculosis 1430BH     1430BH   \n",
       "4    1438842.3                  Mycobacterium tuberculosis 1755BN     1755BN   \n",
       "..         ...                                                ...        ...   \n",
       "704   652616.4  Mycobacterium tuberculosis str. Erdman = ATCC ...     Erdman   \n",
       "705   1773.199        Mycobacterium tuberculosis strain MAL010075  MAL010075   \n",
       "706   1773.200        Mycobacterium tuberculosis strain MAL010077  MAL010077   \n",
       "708   1773.202        Mycobacterium tuberculosis strain MAL020139  MAL020139   \n",
       "709   1773.203        Mycobacterium tuberculosis strain MAL020151  MAL020151   \n",
       "\n",
       "    taxon_id genome_length gc_content fine_consistency checkm_completeness  \\\n",
       "0    1438833       4406079       65.6             99.3                 100   \n",
       "1    1438835       4383014       65.6             99.2                 100   \n",
       "2    1438837       4417210       65.6             99.3                 100   \n",
       "3    1438838       4418670       65.6             99.3                 100   \n",
       "4    1438842       4394400       65.6             99.3                 100   \n",
       "..       ...           ...        ...              ...                 ...   \n",
       "704   652616       4417384      65.65             99.3                 100   \n",
       "705     1773       4421633       65.6             99.3                 100   \n",
       "706     1773       4421308       65.6             99.3                 100   \n",
       "708     1773       4413724       65.6             99.4                 100   \n",
       "709     1773       4396046       65.6             99.3                 100   \n",
       "\n",
       "    checkm_contamination       completion_date  ... sequences patric_cds  \\\n",
       "0                      0  2014-04-29T00:00:00Z  ...       NaN       4372   \n",
       "1                      0  2014-04-29T00:00:00Z  ...       NaN       4333   \n",
       "2                      0  2014-04-29T00:00:00Z  ...       NaN       4384   \n",
       "3                      0  2014-04-29T00:00:00Z  ...       NaN       4383   \n",
       "4                      0  2014-04-29T00:00:00Z  ...       NaN       4360   \n",
       "..                   ...                   ...  ...       ...        ...   \n",
       "704                    0                   NaN  ...       NaN       4372   \n",
       "705                    0  2014-04-29T00:00:00Z  ...       NaN       4320   \n",
       "706                    0  2014-04-29T00:00:00Z  ...       NaN       4304   \n",
       "708                    0  2014-04-29T00:00:00Z  ...       NaN       4307   \n",
       "709                    0  2014-04-29T00:00:00Z  ...       NaN       4312   \n",
       "\n",
       "    refseq_cds isolation_site isolation_source isolation_comments  \\\n",
       "0         4006            NaN              neg                NaN   \n",
       "1         3982            NaN     Bodily fluid                NaN   \n",
       "2         4022            NaN     Bodily fluid                NaN   \n",
       "3         4023            NaN     Bodily fluid                NaN   \n",
       "4         3987            NaN     Bodily fluid                NaN   \n",
       "..         ...            ...              ...                ...   \n",
       "704       4006            NaN              NaN                NaN   \n",
       "705       4006            NaN              NaN                NaN   \n",
       "706       4006            NaN              NaN                NaN   \n",
       "708       4003            NaN              NaN                NaN   \n",
       "709       4019            NaN              NaN                NaN   \n",
       "\n",
       "    collection_date geographic_location            host_name          comments  \n",
       "0              2013             Romania  Human, Homo sapiens  TB-ARC - Romania  \n",
       "1              2012             Romania  Human, Homo sapiens  TB-ARC - Romania  \n",
       "2              2013             Romania  Human, Homo sapiens  TB-ARC - Romania  \n",
       "3              2013             Romania  Human, Homo sapiens  TB-ARC - Romania  \n",
       "4              2013             Romania  Human, Homo sapiens  TB-ARC - Romania  \n",
       "..              ...                 ...                  ...               ...  \n",
       "704             NaN                 NaN                  NaN               NaN  \n",
       "705            2007                Mali  Human, Homo sapiens     TB-ARC - Mali  \n",
       "706             NaN                Mali  Human, Homo sapiens     TB-ARC - Mali  \n",
       "708            2007                Mali  Human, Homo sapiens     TB-ARC - Mali  \n",
       "709            2007                Mali  Human, Homo sapiens     TB-ARC - Mali  \n",
       "\n",
       "[668 rows x 33 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GENOMES_BEFORE_MASH = path.join(DATA_DIR, 'genomes_before_mash.xlsx')\n",
    "df_genomes = pd.read_excel(GENOMES_BEFORE_MASH, dtype=str)\n",
    "\n",
    "df_genomes = df_genomes[df_genomes.genome_id.isin(genes.genome_id)]\n",
    "\n",
    "GENOMES_READY = path.join(DATA_DIR, 'genomes_ready.xlsx')\n",
    "df_genomes.to_excel(GENOMES_READY, index=False)\n",
    "\n",
    "df_genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "216478e7-9886-4ecc-87e1-fb9afdab5844",
   "metadata": {},
   "source": [
    "### 2.2 Panaroo"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c8ccfa5-c026-427e-b28c-1ec8e789f0c6",
   "metadata": {},
   "source": [
    "**INGAT** tambahkan genome reference lagi!\n",
    "\n",
    "- Genome ID: **83332.12**\n",
    "- Genome Name: **Mycobacterium tuberculosis H37Rv**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c3b7f5ee-36aa-472a-a503-d8f77741752e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "reference_id = '83332.12'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "107c5d31-b83c-47c9-a20a-59d26690dc0e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ID_LIST_TXT = path.join(SAMPLES_DIR, '_id_ready.txt')\n",
    "with open(ID_LIST_TXT, 'w') as f:\n",
    "    gids = [reference_id]\n",
    "    gids.extend(df_amr.genome_id.unique())\n",
    "    for gid in gids:\n",
    "        f.write(f'{gid}\\n')\n",
    "\n",
    "selected_antibiotics = df_amr.antibiotic.unique()\n",
    "for antibiotic in selected_antibiotics:    \n",
    "    PANAROO_LIST_TXT = path.join(SAMPLES_DIR, f'_panaroo_{antibiotic}_ready.txt')\n",
    "    gids = [reference_id]\n",
    "    gids.extend(df_amr[df_amr.antibiotic == antibiotic]['genome_id'].to_list())\n",
    "    with open(PANAROO_LIST_TXT, 'w') as f:\n",
    "        for gid in gids:\n",
    "            f.write(f'{gid}.PATRIC.gff {gid}.fna\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "93048cde-fa5b-4b6a-b415-bb6f7e400f31",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>resistant</th>\n",
       "      <th>susceptible</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rifampin</th>\n",
       "      <td>179</td>\n",
       "      <td>201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ethambutol</th>\n",
       "      <td>160</td>\n",
       "      <td>89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>isoniazid</th>\n",
       "      <td>406</td>\n",
       "      <td>175</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            resistant  susceptible\n",
       "rifampin          179          201\n",
       "ethambutol        160           89\n",
       "isoniazid         406          175"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Plot the distributions\n",
    "dist = defaultdict(int)\n",
    "for d in df_amr[['antibiotic', 'phenotype']].itertuples():\n",
    "    name = f'{d.antibiotic}:{d.phenotype}'\n",
    "    dist[name] += 1\n",
    "    \n",
    "antibiotics = set([a.split(':')[0] for a in dist.keys()])\n",
    "counts = [(dist[f'{a}:R'], dist[f'{a}:S']) for a in antibiotics]\n",
    "pd.DataFrame(counts, index=list(antibiotics), columns=['resistant', 'susceptible'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82aeae9d-508c-4e8c-9aad-be441a25215d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
